internal/lifecycle/lifecycle.go

```diff
@@ -74,18 +103,92 @@ func (l *Lifecycle) Start(ctx context.Context) error {
 // hooks run in reverse order.
 func (l *Lifecycle) Stop(ctx context.Context) error {
 	var errs []error
+	l.records = make(HookRecords, 0, l.numStarted)
```

Comment:
I slightly dislike that the start and stop records are on the same field, so the Stop records will overwrite the Start records, but I don't foresee any issues with this in the near-term because if Start was successful, we won't need the records anymore -- at least not in the current iteration of this change.

We may want to do this slightly differently once we're emitting this information to the event logger.